In [1]:
import sys
sys.path.append("..")

In [2]:
from src.load_data import load_tables

In [3]:
tables = load_tables()

In [4]:
users = tables['users']
drivers = tables['drivers']
riders = tables['riders']
locations = tables['locations']
trips = tables['trips']
payments = tables['payments']
reviews = tables['reviews']
cancellations = tables['cancellations']

## UNDERSTANDING THE DATA

### USERS TABLE
- Number of Rows : 2000
- Number of Columns : 7
- Missing Values : None
- Duplicate Rows : None

In [5]:
users.shape

(2000, 7)

In [6]:
users.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   user_id      2000 non-null   int64
 1   name         2000 non-null   str  
 2   email        2000 non-null   str  
 3   phone        2000 non-null   str  
 4   city         2000 non-null   str  
 5   date_joined  2000 non-null   str  
 6   is_driver    2000 non-null   int64
dtypes: int64(2), str(5)
memory usage: 109.5 KB


In [7]:
users.describe(include='all')

,user_id,name,email,phone,city,date_joined,is_driver
count,2000.000000,2000,2000,2000,2000,2000,2000.0000
unique,NaN,1764,2000,2000,4,1243,NaN
top,NaN,Diane Hughes,david.white@gmail.com,+1-809-169-4853,Houston,2021-12-07,NaN
freq,NaN,3,1,1,552,7,NaN
mean,1000.500000,NaN,NaN,NaN,NaN,NaN,0.2000
std,577.494589,NaN,NaN,NaN,NaN,NaN,0.4001
min,1.000000,NaN,NaN,NaN,NaN,NaN,0.0000
25%,500.750000,NaN,NaN,NaN,NaN,NaN,0.0000
50%,1000.500000,NaN,NaN,NaN,NaN,NaN,0.0000
75%,1500.250000,NaN,NaN,NaN,NaN,NaN,0.0000


In [8]:
users.isnull().sum()

user_id        0
name           0
email          0
phone          0
city           0
date_joined    0
is_driver      0
dtype: int64

In [9]:
users.duplicated().sum()

np.int64(0)

### DRIVERS TABLE
- Number of Rows : 400
- Number of Columns : 9
- Missing Values : None
- Duplicate Rows : None

In [10]:
drivers.shape

(400, 9)

In [11]:
drivers.info()

<class 'pandas.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   driver_id      400 non-null    int64  
 1   user_id        400 non-null    int64  
 2   vehicle_make   400 non-null    str    
 3   vehicle_model  400 non-null    str    
 4   vehicle_year   400 non-null    int64  
 5   license_plate  400 non-null    str    
 6   rating         400 non-null    float64
 7   join_date      400 non-null    str    
 8   is_active      400 non-null    int64  
dtypes: float64(1), int64(4), str(4)
memory usage: 28.3 KB


In [12]:
drivers.describe(include='all')

,driver_id,user_id,vehicle_make,vehicle_model,vehicle_year,license_plate,rating,join_date,is_active
count,400.000000,400.00000,400,400,400.000000,400,400.000000,400,400.000000
unique,NaN,NaN,10,50,NaN,400,NaN,348,NaN
top,NaN,NaN,Nissan,Equinox,NaN,ZVJ-2140,NaN,2019-12-19,NaN
freq,NaN,NaN,50,17,NaN,1,NaN,3,NaN
mean,200.500000,995.78500,NaN,NaN,2017.530000,NaN,4.152575,NaN,0.877500
std,115.614301,588.38962,NaN,NaN,3.521378,NaN,0.532396,NaN,0.328273
min,1.000000,2.00000,NaN,NaN,2012.000000,NaN,2.590000,NaN,0.000000
25%,100.750000,468.50000,NaN,NaN,2014.000000,NaN,3.770000,NaN,1.000000
50%,200.500000,1017.00000,NaN,NaN,2018.000000,NaN,4.180000,NaN,1.000000
75%,300.250000,1497.50000,NaN,NaN,2020.250000,NaN,4.610000,NaN,1.000000


In [13]:
drivers.isnull().sum()

driver_id        0
user_id          0
vehicle_make     0
vehicle_model    0
vehicle_year     0
license_plate    0
rating           0
join_date        0
is_active        0
dtype: int64

In [14]:
drivers.duplicated().sum()

np.int64(0)

### RIDERS TABLE
- Number of Rows : 1600
- Number of Columns : 5
- Missing Values : None
- Duplicate Rows : None

In [15]:
riders.shape

(1600, 5)

In [16]:
riders.info()

<class 'pandas.DataFrame'>
RangeIndex: 1600 entries, 0 to 1599
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   rider_id     1600 non-null   int64  
 1   user_id      1600 non-null   int64  
 2   rating       1600 non-null   float64
 3   total_trips  1600 non-null   int64  
 4   created_at   1600 non-null   str    
dtypes: float64(1), int64(3), str(1)
memory usage: 62.6 KB


In [17]:
riders.describe(include='all')

,rider_id,user_id,rating,total_trips,created_at
count,1600.00000,1600.000000,1600.000000,1600.000000,1600
unique,NaN,NaN,NaN,NaN,1600
top,NaN,NaN,NaN,NaN,2020-09-13 01:52:12
freq,NaN,NaN,NaN,NaN,1
mean,800.50000,1001.678750,3.991137,10.516875,NaN
std,462.02453,574.919114,0.590380,17.336344,NaN
min,1.00000,1.000000,3.000000,0.000000,NaN
25%,400.75000,515.500000,3.450000,1.000000,NaN
50%,800.50000,998.500000,3.990000,3.000000,NaN
75%,1200.25000,1501.250000,4.500000,13.000000,NaN


In [18]:
riders.isnull().sum()

rider_id       0
user_id        0
rating         0
total_trips    0
created_at     0
dtype: int64

In [19]:
riders.duplicated().sum()

np.int64(0)

### LOCATIONS TABLE
- Number of rows : 40
- Number of columns : 6
- Missing Values : None
- Duplicate rows : None

In [20]:
locations.shape

(40, 6)

In [21]:
locations.info()

<class 'pandas.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   location_id  40 non-null     int64  
 1   zone_name    40 non-null     str    
 2   city         40 non-null     str    
 3   latitude     40 non-null     float64
 4   longitude    40 non-null     float64
 5   zone_type    40 non-null     str    
dtypes: float64(2), int64(1), str(3)
memory usage: 2.0 KB


In [22]:
locations.describe(include='all')

,location_id,zone_name,city,latitude,longitude,zone_type
count,40.000000,40,40,40.000000,40.000000,40
unique,NaN,40,4,NaN,NaN,4
top,NaN,Times Square,New York,NaN,NaN,residential
freq,NaN,1,10,NaN,NaN,16
mean,20.500000,NaN,NaN,36.606307,-93.843383,NaN
std,11.690452,NaN,NaN,5.028859,16.302392,NaN
min,1.000000,NaN,NaN,29.619700,-118.491000,NaN
25%,10.750000,NaN,NaN,32.919400,-101.262175,NaN
50%,20.500000,NaN,NaN,37.411050,-91.593100,NaN
75%,30.250000,NaN,NaN,41.055400,-84.193700,NaN


In [23]:
locations.isnull().sum()

location_id    0
zone_name      0
city           0
latitude       0
longitude      0
zone_type      0
dtype: int64

In [24]:
locations.duplicated().sum()

np.int64(0)

### TRIPS TABLE
- Number of rows : 2000
- Number of columns : 15
- Missing values : Yes 
- Duplicate rows : None

In [25]:
trips.shape

(20000, 15)

In [26]:
trips.info()

<class 'pandas.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   trip_id              20000 non-null  int64  
 1   rider_id             20000 non-null  int64  
 2   driver_id            20000 non-null  int64  
 3   pickup_location_id   20000 non-null  int64  
 4   dropoff_location_id  20000 non-null  int64  
 5   requested_at         20000 non-null  str    
 6   started_at           20000 non-null  str    
 7   completed_at         16827 non-null  str    
 8   status               20000 non-null  str    
 9   distance_km          20000 non-null  float64
 10  duration_mins        20000 non-null  int64  
 11  base_fare            20000 non-null  float64
 12  surge_multiplier     20000 non-null  float64
 13  total_fare           20000 non-null  float64
 14  payment_method       20000 non-null  str    
dtypes: float64(4), int64(6), str(5)
memory usage: 2

In [27]:
trips.describe(include='all')

,trip_id,rider_id,driver_id,pickup_location_id,dropoff_location_id,requested_at,started_at,completed_at,status,distance_km,duration_mins,base_fare,surge_multiplier,total_fare,payment_method
count,20000.000000,20000.00000,20000.000000,20000.000000,20000.000000,20000,20000,16827,20000,20000.000000,20000.000000,20000.000000,20000.000000,20000.000000,20000
unique,NaN,NaN,NaN,NaN,NaN,19997,19998,16825,3,NaN,NaN,NaN,NaN,NaN,3
top,NaN,NaN,NaN,NaN,NaN,2024-04-11 09:18:47,2023-06-22 07:06:27,2023-11-14 22:36:24,completed,NaN,NaN,NaN,NaN,NaN,card
freq,NaN,NaN,NaN,NaN,NaN,2,2,2,16827,NaN,NaN,NaN,NaN,NaN,6741
mean,10000.500000,808.04035,201.431950,20.123100,20.116850,NaN,NaN,NaN,NaN,18.013124,31.318850,28.578512,1.261648,36.008345,NaN
std,5773.647028,458.81708,114.642109,11.951405,11.998183,NaN,NaN,NaN,NaN,11.383895,20.164943,15.886229,0.533862,26.526153,NaN
min,1.000000,1.00000,1.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,0.500000,3.000000,3.650000,1.000000,3.650000,NaN
25%,5000.750000,414.00000,101.000000,9.000000,9.000000,NaN,NaN,NaN,NaN,8.060000,14.000000,14.170000,1.000000,16.210000,NaN
50%,10000.500000,801.00000,207.000000,20.000000,20.000000,NaN,NaN,NaN,NaN,17.800000,30.000000,29.495000,1.000000,32.470000,NaN
75%,15000.250000,1226.00000,298.000000,31.000000,31.000000,NaN,NaN,NaN,NaN,25.340000,46.000000,39.642500,1.000000,45.242500,NaN


In [28]:
trips.isnull().sum()

trip_id                   0
rider_id                  0
driver_id                 0
pickup_location_id        0
dropoff_location_id       0
requested_at              0
started_at                0
completed_at           3173
status                    0
distance_km               0
duration_mins             0
base_fare                 0
surge_multiplier          0
total_fare                0
payment_method            0
dtype: int64

In [29]:
trips.duplicated().sum()

np.int64(0)

### PAYMENTS TABLE
- Number of rows : 16827
- Number of columns : 6
- Missing Values : None
- Duplicate Rows : None

In [30]:
payments.shape

(16827, 6)

In [31]:
payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 16827 entries, 0 to 16826
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   payment_id  16827 non-null  int64  
 1   trip_id     16827 non-null  int64  
 2   amount      16827 non-null  float64
 3   method      16827 non-null  str    
 4   status      16827 non-null  str    
 5   paid_at     16827 non-null  str    
dtypes: float64(1), int64(2), str(3)
memory usage: 788.9 KB


In [32]:
payments.describe(include='all')

,payment_id,trip_id,amount,method,status,paid_at
count,16827.000000,16827.000000,16827.000000,16827,16827,16827
unique,NaN,NaN,NaN,3,3,16825
top,NaN,NaN,NaN,card,success,2023-03-28 20:53:07
freq,NaN,NaN,NaN,5700,16359,2
mean,8414.000000,10011.081714,35.929751,NaN,NaN,NaN
std,4857.680825,5779.965539,26.441414,NaN,NaN,NaN
min,1.000000,1.000000,3.650000,NaN,NaN,NaN
25%,4207.500000,5016.500000,16.200000,NaN,NaN,NaN
50%,8414.000000,10022.000000,32.490000,NaN,NaN,NaN
75%,12620.500000,15025.500000,45.120000,NaN,NaN,NaN


In [33]:
payments.isnull().sum()

payment_id    0
trip_id       0
amount        0
method        0
status        0
paid_at       0
dtype: int64

In [34]:
payments.duplicated().sum()

np.int64(0)

### REVIEWS TABLE
- Number of rows : 15136
- Number of columns : 7
- Missing values : Yes
- Duplicate Rows : None

In [35]:
reviews.shape

(15136, 7)

In [36]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 15136 entries, 0 to 15135
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   review_id    15136 non-null  int64
 1   trip_id      15136 non-null  int64
 2   reviewer_id  15136 non-null  int64
 3   reviewee_id  15136 non-null  int64
 4   rating       15136 non-null  int64
 5   comment      9216 non-null   str  
 6   reviewed_at  15136 non-null  str  
dtypes: int64(5), str(2)
memory usage: 827.9 KB


In [37]:
reviews.describe(include='all')

,review_id,trip_id,reviewer_id,reviewee_id,rating,comment,reviewed_at
count,15136.000000,15136.000000,15136.000000,15136.000000,15136.000000,9216,15136
unique,NaN,NaN,NaN,NaN,NaN,31,15126
top,NaN,NaN,NaN,NaN,NaN,Super friendly and helpful.,2023-10-19 21:36:08
freq,NaN,NaN,NaN,NaN,NaN,1091,2
mean,7568.500000,10016.379757,1009.110663,999.172569,3.654070,NaN,NaN
std,4369.531172,5778.288510,576.700326,581.698296,0.869887,NaN,NaN
min,1.000000,1.000000,1.000000,1.000000,1.000000,NaN,NaN
25%,3784.750000,5012.500000,510.000000,485.000000,3.000000,NaN,NaN
50%,7568.500000,10057.000000,1044.000000,1028.500000,4.000000,NaN,NaN
75%,11352.250000,15016.250000,1489.000000,1483.000000,4.000000,NaN,NaN


In [38]:
reviews.isnull().sum()

review_id         0
trip_id           0
reviewer_id       0
reviewee_id       0
rating            0
comment        5920
reviewed_at       0
dtype: int64

In [39]:
reviews.duplicated().sum()

np.int64(0)

### CANCELLATIONS TABLE
- Number of rows : 2966
- Number of columns : 5
- Missing Values : None
- Duplicate rows : None

In [40]:
cancellations.shape

(2966, 5)

In [41]:
cancellations.info()

<class 'pandas.DataFrame'>
RangeIndex: 2966 entries, 0 to 2965
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   cancel_id     2966 non-null   int64
 1   trip_id       2966 non-null   int64
 2   cancelled_by  2966 non-null   str  
 3   reason        2966 non-null   str  
 4   cancelled_at  2966 non-null   str  
dtypes: int64(2), str(3)
memory usage: 116.0 KB


In [42]:
cancellations.describe(include='all')

,cancel_id,trip_id,cancelled_by,reason,cancelled_at
count,2966.000000,2966.000000,2966,2966,2966
unique,NaN,NaN,2,14,2966
top,NaN,NaN,rider,personal emergency,2024-01-25 20:55:28
freq,NaN,NaN,2069,396,1
mean,1483.500000,9950.007417,NaN,NaN,NaN
std,856.354775,5734.265379,NaN,NaN,NaN
min,1.000000,2.000000,NaN,NaN,NaN
25%,742.250000,4942.750000,NaN,NaN,NaN
50%,1483.500000,9903.500000,NaN,NaN,NaN
75%,2224.750000,14947.750000,NaN,NaN,NaN


In [43]:
cancellations.isnull().sum()

cancel_id       0
trip_id         0
cancelled_by    0
reason          0
cancelled_at    0
dtype: int64

In [44]:
cancellations.duplicated().sum()

np.int64(0)